core/runtime.rs

```diff
@@ -748,20 +852,30 @@ impl JsRuntime {
         let promise_id = promise.get_identity_hash();
         let mut state = state_rc.borrow_mut();
         state.pending_promise_exceptions.remove(&promise_id);
+        let promise_global = v8::Global::new(scope, promise);
+        state
+          .pending_mod_evaluate
+          .insert(id, (promise_global, sender));
       } else {
         assert!(status == v8::ModuleStatus::Errored);
       }
     }
 
-    match status {
-      v8::ModuleStatus::Evaluated => Ok(()),
-      v8::ModuleStatus::Errored => {
-        let exception = module.get_exception();
-        exception_to_err_result(scope, exception)
-          .map_err(|err| attach_handle_to_error(scope, err, exception))
+    Ok(receiver)
+  }
+
+  pub async fn mod_evaluate(&mut self, id: ModuleId) -> Result<(), AnyError> {
```

Comment:
Is it odd that JsRuntime itself is a future, but it now has a method which also returns a future?

Reply:
@ry there is a precedent with `JsRuntime::load_module` which we've had for a couple months.

Reply:
I guess what happens if you try to `tokio::select!` on both the result of `mod_evaluate` and the JsRuntime itself?

Reply:
`JsRuntime` would be polled twice (because it's polled once in `mod_evaluate`). I don't like `JsRuntime` being a future, there's only a single time that you want to run event loop to complication - in `main.rs`. I would prefer to have explicit `JsRuntime::poll_event_loop()` instead.

Reply:
> What happens if you try to `tokio::select!` on both the result of `mod_evaluate` and the JsRuntime itself?

That either isn't or shouldn't be possible. 
You need a &mut reference to a Future to poll it, so you can't poll X and a container with X in it at the same time.

Reply:
Actually @piscisaureus is right:
```rust
  let fut = self.isolate.mod_evaluate(id);
    tokio::select! {
      _ = fut => {
        // 
        Ok(())
      }
      _ = self.isolate => {
        //
        Ok(())
      }
    }
```

```
error[E0507]: cannot move out of `self.isolate` which is behind a mutable reference
   --> cli/worker.rs:198:11
    |
198 |       _ = self.isolate => {
    |           ^^^^^^^^^^^^ move occurs because `self.isolate` has type `deno_core::JsRuntime`, which does not implement the `Copy` trait

error[E0505]: cannot move out of `self.isolate` because it is borrowed
   --> cli/worker.rs:198:11
    |
192 |       let fut = self.isolate.mod_evaluate(id);
    |                 ------------ borrow of `self.isolate` occurs here
193 | /     tokio::select! {
194 | |       _ = fut => {
195 | |         //
196 | |         Ok(())
197 | |       }
198 | |       _ = self.isolate => {
    | |           ^^^^^^^^^^^^ move out of `self.isolate` occurs here
...   |
201 | |       }
202 | |     }
    | |_____- borrow later used here
```